### DISCUSSION TITLE
Is it possible to "send all" with SingleWalletAddress ?

### DISCUSSION THREAD

**Discussion by pitersky** [input-output-hk/cardano-js-sdk#425] [Q&A] [answered]
Hi!

Is it possible to construct a transaction without change output using SingleWalletAddress instance? I am trying to do a "Send All" feature (I am not sending any tokens, only ADA). So far I can build and send "normal" transactions, but when I try to use all available coins in `initializeTx` the `InputSelectionError: UTxO Fully Depleted` error raises.

The code I use looks like

```js
const available = await firstValueFrom(singleWalletInstance.balance.utxo.available$)
const address = 'any_address'

const tx = await singleWalletInstance.initializeTx({
  outputs: new Set([{ address, value: { coins: available.coins } }]),
})
```

**Comment by mkazlauskas** [comment, marked as answer]
Hi, unfortunately this use case is currently unsupported. We're planning to add it in the future, but it's currently low priority.

This feature is more complicated than it might look as
- there is an edge case where it can only be done with multiple transactions (could exceed max transaction size)
- preserving utxo distribution is desired

In your case (sending coin only, no assets) it is not as complicated. You could try to re-implement `initializeTx` by referencing SDK code and replacing the call to [input selection](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/wallet/src/SingleAddressWallet.ts#L370) with a simpler implementation that selects all utxo and balances the output(s) with fee in a a way that doesn't need any change.

**Reply by pitersky** [reply]
Thank you for your reply! So I am detecting whether a user willing to send all available and returning a predefined "shape" of selecting without a change. It is now looks like

```js
export function createInputSelector () {
  const inputSelector = roundRobinRandomImprove()

  const select = inputSelector.select

  const selectAll = async (...args) => {
    const { constraints: { computeMinimumCost }, utxo: utxos, outputs } = args[0]

    if (outputs.size !== 1) {
      throw new Error('must contain exactly 1 output')
    }

    let all = 0n

    for (const utxo of utxos.values()) {
      all += utxo[1].value.coins
    }

    let address
    let sending

    for (const output of outputs.values()) {
      address = output.address
      sending = output.value.coins
    }

    const fee = BigInt(await computeMinimumCost({
      change: new Set(),
      fee: cslUtil.MAX_U64,
      inputs: new Set(utxos),
      outputs,
    }))

    const available = all - fee

    // if not sendAll then fallback to normal select
    if (sending !== available && sending !== all) {
      return select(...args)
    }

    return {
      remainingUTxO: new Set(), // all utxos are used as inputs, no utxo remaining
      selection: {
        change: [], // no change output
        fee,
        inputs: utxos, // use all utxos as inputs
        outputs: new Set([{ address, value: { coins: available } }]),
      },
    }
  }

  inputSelector.select = selectAll

  return inputSelector
}
```

It has that part
```
    if (sending !== available && sending !== all) {
      return select(...args)
    }
```
since we are not actually showing a user all available coins, but only those that can be received by the other side, which is available minus a fee. But it seems to work for now.

And than that is goes to the SingleWalletAddress instantiation as a custom inputSelector

**Reply by mkazlauskas** [reply]
Good idea to inject it as a custom input selector! Thank you for sharing, could be useful for someone else too.

A small improvement you could make is using `Cardano.util.coalesceValueQuantities` instead of `all += utxo[1].value.coins` - it also adds assets if present.
